# BYTE PAIR ENCODING

<div class="alert alert-block alert-success">
We implemented a simple tokenization scheme. We took sentences, converted them into tokens and then we converted this into
vocabulary so then every word of the token was arranged in an ascending order and then a ID or a numerical ID or a token ID was 
assigned to each token so here every word was a unique token along with special characters like comma full stop exclamation mark
etc.
</div>

<div class="alert alert-block alert-info">
The tokenization algorithms are of three types <br>
* word based tokenizer <br>
* subword based tokenizer <br>
* character based tokenizer <br>
</div>

## Word Based tokenizer
Consider the example "the fox chased the dog," where the tokens are "the," "fox," "chased," "the," and "dog." Each word serves as an individual token, illustrating a word-based tokenizer. In this approach, tokens are defined at word boundaries.

However, there are challenges associated with word-based tokenizers. The primary issue arises when dealing with words that are not included in the vocabulary. Typically, you have a large dataset that you break down into sentences, and then into individual words, assigning a token ID to each. But when a user interacts with a language model and inputs a word not found in the vocabulary-referred to as "out of vocabulary" (OOV) words-the tokenization process often encounters errors. Handling OOV words is notoriously difficult with word-based tokenization.

A vocabulary is essentially a dictionary of tokens organized in ascending order, where each token corresponds to a unique token ID. Another issue with this tokenization method is exemplified by the words "boy" and "boys." Although they are similar, they are treated as separate tokens, with potentially very different token IDs, depending on their position in the vocabulary. This approach fails to capture the semantic similarity between such words, as "boy" is the root word for "boys," yet both are treated distinctly.

Additionally, attempting to account for all possible word combinations is not scalable, as it significantly enlarges the vocabulary size. These are the main challenges associated with word-based tokenization.

## Character based tokenizer
On the opposite end of the spectrum is character-based tokenization, another widely used tokenization scheme. Instead of using whole words as tokens, this approach treats individual characters as tokens. For instance, in the sentence "my hobby is playing Cricket," each character becomes a token. Ignoring spaces, the first few tokens would be 'm', 'y', 'h', 'o', 'b', and so on.

This method results in a very small vocabulary size, as every language has a limited number of characters. In English, there are only 256 characters, compared to around 170,000 to 200,000 words. One of the significant advantages of character-based tokenization is the elimination of the "out of vocabulary" (OOV) problem. Any new sentence can be broken down into characters, all of which will belong to the existing set of 256 characters in the vocabulary, thus avoiding OOV issues. This also means that the vocabulary size is manageable, unlike word-based tokenization, which requires a massive vocabulary to cover all possible words.

While character-based tokenization offers solutions like OOV problem elimination and computational efficiency due to its small vocabulary size, it also has drawbacks. The primary issue is the loss of semantic meaning associated with words. Language models rely on the meaning of words, and by breaking sentences into individual characters, this meaning is lost. For example, related words like "boy" and "boys" share a common meaning, but this connection is lost in character-level tokenization.

Another downside is that the tokenized sequence becomes much longer than the original text. For example, the word "dinosaur" would be treated as one token in word-based tokenization but split into eight separate tokens ('d', 'i', 'n', 'o', 's', 'a', 'u', 'r') in character-based tokenization. This results in a longer tokenized sequence, which can be cumbersome to handle.

In summary, while word-based tokenization struggles with large vocabulary sizes and OOV words, it preserves semantic meaning. Character-based tokenization resolves the OOV issue and maintains a smaller vocabulary but sacrifices word meaning and increases tokenized sequence length.

## Subword Tokenizer

Let's explore subword tokenization, a method that combines elements of both word and character tokenization. Byte Pair Encoding (BPE) is an example of this approach. Subword tokenization effectively captures root words found in various forms, like "boy" in both "boy" and "boys." Here's how it works:

Rule 1: For words that appear frequently in the dataset, do not split them into smaller subwords. These words should be retained as they are.

Rule 2: For rare words that don't appear often, split them into smaller, meaningful subwords. This rule is crucial because it allows rare words to be broken down, even to the character level if necessary.

This approach is a blend of word and character tokenization. The first rule keeps common words intact, similar to word tokenization. The second rule allows for further breakdown of rare words, borrowing from character tokenization.

For example, if "boy" appears frequently in the dataset, it remains as a single token, following Rule 1. However, if "boys" is encountered less often, it can be split into "boy" and "s," as "boys" derives from the root "boy."

Subword tokenization offers several advantages. Firstly, it helps models recognize that words with the same root, like "token," "tokens," and "tokenizing," share a common meaning, which is often lost in word-based and character-based tokenization. Secondly, it aids in understanding patterns, such as the shared suffix in "tokenization" and "modernization." The suffix "zation" can be treated as a subword, highlighting similar syntactic roles.

Subword tokenization is significant because it allows for more nuanced word splitting. Byte Pair Encoding (BPE) is a subword tokenization algorithm used by modern language models like GPT-2 and GPT-3.

**BPE Tokenizer**

In [1]:
#! pip3 install tiktoken

<div class="alert alert-block alert-success">
tiktoken is the python library that implements BPE
</div>

In [2]:
import importlib
import tiktoken

print("tiktoken version:", importlib.metadata.version("tiktoken"))

tiktoken version: 0.7.0


<div class="alert alert-block alert-success">
Instantiate the byte pair tokenizer
and encode and decode a text
</div>

In [3]:
tokenizer = tiktoken.get_encoding("gpt2")

In [4]:
text = (
    "Hello, do you like tea? <|endoftext|> In the sunlit terraces"
     "of someunknownPlace."
)

integers = tokenizer.encode(text, allowed_special={"<|endoftext|>"})

print(integers)

[15496, 11, 466, 345, 588, 8887, 30, 220, 50256, 554, 262, 4252, 18250, 8812, 2114, 1659, 617, 34680, 27271, 13]


<div class="alert alert-block alert-success">
Note the tokenId 50256 - it is the "endoftext" token - also shows the vocabulary size of the toeknizer - much less than 170k to 200k tokens in English Language
</div>

In [5]:
strings = tokenizer.decode(integers)

print(strings)

Hello, do you like tea? <|endoftext|> In the sunlit terracesof someunknownPlace.


<div class="alert alert-block alert-success">
Byte Pair Encoding offers several benefits, particularly in handling out-of-vocabulary words. Firstly, it helps reduce the vocabulary size. Secondly, it effectively manages unfamiliar text.
It can handle unfamiliar text by breaking down words not found in its predefined vocabulary into smaller subword units or even individual characters. This is clearly seen in example below - for this unknown word."
</div>

In [8]:
integers = tokenizer.encode("qytzrk")
print(integers)

strings = tokenizer.decode(integers)
print(strings)

[80, 20760, 89, 81, 74]
qytzrk


**Data sampling with sliding window**

In [7]:
with open("the-verdict.txt", "r", encoding="utf-8") as f:
    raw_text = f.read()

enc_text = tokenizer.encode(raw_text)
print(len(enc_text))

5775


In [8]:
enc_sample = enc_text[50:]


In [9]:
context_size = 4

x = enc_sample[:context_size]
y = enc_sample[1:context_size+1]

print(f"x: {x}")
print(f"y:      {y}")

x: [11, 339, 550, 5710]
y:      [339, 550, 5710, 465]


In [10]:
for i in range(1, context_size+1):
    context = enc_sample[:i]
    desired = enc_sample[i]

    print(context, "---->", desired)

[11] ----> 339
[11, 339] ----> 550
[11, 339, 550] ----> 5710
[11, 339, 550, 5710] ----> 465


In [11]:
for i in range(1, context_size+1):
    context = enc_sample[:i]
    desired = enc_sample[i]

    print(tokenizer.decode(context), "---->", tokenizer.decode([desired]))

, ---->  he
, he ---->  had
, he had ---->  dropped
, he had dropped ---->  his
